### PR TITLE
Fix/misc

### PR DESCRIPTION
**Pull request by boillodmanuel**
Several fixes related to https://github.com/swagger-api/swagger-parser/issues/398.

I did one commit per fix (see commit message) to facilitate the review

I'll be happy to discuss each commit
- commit `#e33949d` with message `feat(resolve): add settings to allow resolving references without adding resource parameters to each operation` do not change public behaviour (of `SwaggerParser`) but for those using directly `SwaggerResolver`, it allow to don't add resource parameters on operation, and have a behaviour closer to `ObjectMapper` and also to keep the `PathParameter`on path (not flatten to each operation)

### CODE REVIEW

File: modules/swagger-parser/src/main/java/io/swagger/parser/ResolverCache.java [hunk 1]
```diff
@@ -36,14 +36,14 @@
...
-    private Set<String> referencedModelKeys = new HashSet<>();
+    private Map<String, Object> resolutionCache = new LinkedHashMap<>();
+    private Map<String, String> externalFileCache = new LinkedHashMap<>();
+    private Set<String> referencedModelKeys = new LinkedHashSet<>();
 
     /*
     a map that stores original external references, and their associated renamed references
      */
-    private Map<String, String> renameCache = new HashMap<>();
+    private Map<String, String> renameCache = new LinkedHashMap<>();
```
Comment (fehguy) [inline modules/swagger-parser/src/main/java/io/swagger/parser/ResolverCache.java]: i don't think the `LinkedHashMap` is necessary here as order never matters in this cache.  It's just for resolution.

File: modules/swagger-parser/src/main/java/io/swagger/parser/ResolverCache.java [hunk 2]
```diff
@@ -36,14 +36,14 @@
     private final List<AuthorizationValue> auths;
     private final Path parentDirectory;
     private final String rootPath;
-    private Map<String, Object> resolutionCache = new HashMap<>();
-    private Map<String, String> externalFileCache = new HashMap<>();
-    private Set<String> referencedModelKeys = new HashSet<>();
+    private Map<String, Object> resolutionCache = new LinkedHashMap<>();
```
Comment (fehguy) [inline modules/swagger-parser/src/main/java/io/swagger/parser/ResolverCache.java]: HashMap should do fine, did you see an issue with it?

File: modules/swagger-parser/src/main/java/io/swagger/parser/ResolverCache.java [hunk 3]
```diff
@@ -36,14 +36,14 @@
...
+    private Map<String, String> externalFileCache = new LinkedHashMap<>();
```
Comment (fehguy) [inline modules/swagger-parser/src/main/java/io/swagger/parser/ResolverCache.java]: HashMap should do fine, did you see an issue with it? (repeating)

File: modules/swagger-parser/src/main/java/io/swagger/parser/SwaggerResolver.java [hunk 4]
```diff
@@ -53,4 +63,26 @@ public Swagger resolve() {
 
         return swagger;
     }
+
+    public static class Settings {
```
Comment (fehguy) [inline modules/swagger-parser/src/main/java/io/swagger/parser/SwaggerResolver.java]: It's not clear to me why we need this.  The need to add parameters to operations is dictated by the specification and the location of the parameters, not some user-specified setting.

File: modules/swagger-parser/src/main/java/io/swagger/parser/processors/DefinitionsProcessor.java [hunk 5]
```diff
@@ -29,7 +29,7 @@ public void processDefinitions() {
             return;
         }
 
-        Set<String> keySet = new HashSet<>();
+        Set<String> keySet = new LinkedHashSet<>();
```
Comment (fehguy) [inline modules/swagger-parser/src/main/java/io/swagger/parser/processors/DefinitionsProcessor.java]: As with before, not clear why we care about order for these internal storage mechanisms.

File: modules/swagger-parser/src/main/java/io/swagger/parser/processors/PathsProcessor.java [hunk 6]
```diff
@@ -16,12 +17,17 @@
 
     private final Swagger swagger;
     private final ResolverCache cache;
+    private final SwaggerResolver.Settings settings;
```
Comment (fehguy) [inline modules/swagger-parser/src/main/java/io/swagger/parser/processors/PathsProcessor.java]: As with above, I don't see why we need to add a `Settings` here.

File: modules/swagger-parser/src/test/java/io/swagger/parser/SwaggerResolverTest.java [hunk 7]
```diff
@@ -40,7 +41,7 @@ public void testSwaggerResolver(@Injectable final Swagger swagger,
             result = definitionsProcessor;
             times = 1;
 
-            new PathsProcessor(cache, swagger);
+            new PathsProcessor(cache, swagger, withInstanceOf(SwaggerResolver.Settings.class));
```
Comment (fehguy) [inline modules/swagger-parser/src/test/java/io/swagger/parser/SwaggerResolverTest.java]: I don't like this.  What else would be passed in the `PathsProcessor`?  Is this expected to be an extension point?

File: modules/swagger-parser/src/main/java/io/swagger/parser/SwaggerResolver.java (hunk 4 above)
Comment (boillodmanuel) [inline modules/swagger-parser/src/main/java/io/swagger/parser/SwaggerResolver.java]: The last 3 comments is about this `addParametersToEachOperation` setting.
I understand it as it is only useful is specific use cases.

First at all, the variable could have a better name `addResourceParametersToEachOperation`.
The current behaviour (before this PR) is:
- resource parameters are parsed
- when processing operations, resource parameters are merged with operation parameters at operation level and resources parameters are emptied.

This result to a `Swagger` object with no resource parameters.  Even `path variable` are cleaned from resource parameters, which is the kind of parameter which have the most value at resource level. 
In our usage of swagger-parser, we want to keep resource parameters as is.
So the best option I saw to add this capability without changing swagger-parser behaviour is to introduce this settings.

Comment (fehguy) [inline modules/swagger-parser/src/main/java/io/swagger/parser/SwaggerResolver.java]: I see.  So the parser is essentially copying the parameters from the path-item object to the operations "below" it in hierarchy correct?  and what you're saying is, you don't want it to copy them, or you want to copy and leave them there?

Comment (boillodmanuel) [inline modules/swagger-parser/src/main/java/io/swagger/parser/SwaggerResolver.java]: > So the parser is essentially copying the parameters from the path-item object to the operations "below" it in hierarchy correct

Exactly

> you don't want it to copy them, or you want to copy and leave them there?

I don't want to copy them at all.

Before, we were using the swagger-core Object-Mapper, but it didn't support multi-files. So we compare it with swagger-parser and I tried to fix all important difference with this PR.

**Pull request merged**